### ISSUE TITLE
.merge function doesn't fully update file

### ISSUE THREAD

**Issue by fredericlestage-vdm** [canada-ca/ore-ero#614]
I noticed that when using the .merge function of fileWriter with propPath and onValue parameters, only the section specified by these parameters gets updated.

For instance, I could do the following:
fileWriter.merge(file, object, 'releases' ,'name.en');

And that would correctly update the release. However, If I had changed any other value not included in the release data (let's say contact.email), that value wouldn't get updated.

So there's no way to change the core of a code project or a software with the form except when its first created.

I was wondering if that was by design, or is it a flaw we should look into?

**Comment by fredericlestage-vdm** [comment]
After thinking about it, it might have to do only with the new way the data files are separated in my improvement/db branch. So I might be able to find a solution for it by changing git-file-manager.js directly

**Comment by smellems** [comment]
Not sure about the merging exactly, but this sound like it could be true and should be looked at if we are not updating what we think we are..

contact.email is part of release, but if for software if, for example, we updated only 'administrations' and not the other fields defining the software that might have been changed in the form.

**Comment by fredericlestage-vdm** [comment]
More thoughts about this:

Do we want to allow the users to add changes to a software or a standard at all. Because logically, once the software is created, its data shouldn't change that much. I'm trying to define if any of the fields are subject to change. All of the below are assumptions, based on my knowledge:

Software:
 - Name: Once a project has a name, it should't change
 - Description: Might change if the software goes through a new version including a major change
 - Homepage URL: Probably won't change at all, since the softwares are usually hosted on a website relative to its name.
 - Licence: Could maybe change?
    - Also, secondary thought, when does a software have more than one licence?
 - Keywords: Could change, probably if a user sees that it's missing some

Standard:
 - Acronym: Shouldn't change
 - Name: Shouldn't change
 - Description: Probably won't
 - URL for the specification: Shouldn't change either
 - Organization name: Probably won't change, but could
 - Keywords: Same as for software, users may want to add some, but probably won't remove any.

Following these, changes to an administration using the software of standard are more likely, so I would leave the possibility to do so, but once a software of standard is selected, we could disable its fields so they can't be changed. Just to make it clear, I make a distinction between the software and standard fields (the ones listed above) vs the uses or administration fields since in my branch with the new DB they are located within completely different files, and even if we account for the current state of the data, the admin data is already changing and not causing problems, while the base data (for software of standard fields) doesn't necessarily gets updated, as explained in the issue's description).

**Comment by fredericlestage-vdm** [comment]
Another solution would be to also split the forms. 

We could have the 2 sections (software or standard vs use) on separate pages (which has its own set of problems, for instance not being able to create a software and add an administration use at the same time)

Or, when the user select an existing software, don't show its fields directly, but add a button or something that allows it "change the software's data" that would then display the fields (either by enabling them, or by using an accordion => what we currently use with lists/<details>/<summary> for schemas)

Or we could radically change the UX/flow of the form, but that's a whole other issue
![IMG_20190729_161040](https://user-images.githubusercontent.com/50632796/62078882-94186a80-b21b-11e9-93d6-1b6b8b036744.jpg)

**Comment by fredericlestage-vdm** [comment]
The problem actually lays in that if a value for propPath and onValue, it only updates values in that specific array (propPath), and not the level above. So for instance, if you have:
```yaml
name: test
licences:
  - url: 'http://www.mit.com'
    spdxId: MIT
  - url: http://www.apache2.com
    spdxId: Apache-2.0
```
and you call `fileWriter.merge(path, file, 'licences', 'spdxId')`, you will only change (update or add) elements under licences. What that means is that name won't change in any way, and that's the issue I'm facing.

Also, a sub-issue is that there's also no way of removing an element. For instance, if my object `file` doesn't contain a license where the spdxId equals 'MIT', but my original data does, it won't remove it from the original data.

**Comment by fredericlestage-vdm** [comment]
And what if the file has multiple instances of arrays which could individually get updated (like how code form has a partners array, a licences array and a relatedCode array) and so far there's no way of updating them individually using the current merge function.

That's way, maybe the solution could be as simple as overwriting the file each time, which make sure the data in the form is the new data to the file.

**Comment by j-rewerts** [comment]
My intention was that `git-file-manager.js` be as flexible as possible. The behaviour that @fredericlestage-vdm has identified is what I had intended at the time, but I'm now realizing it isn't as useful as I thought it was. This wouldn't work well for your current problem and it also won't work if you want to update multiple elements in separate arrays. Currently, it recreates nested arrays, so if the list of licenses is changed, I think it could show up in GitHub as the entire array of licenses being updated even if it was just 1 that was changed.

Just to clarify something I found confusing after looking at my code, the 4th parameter `onValue` is meant to allow for merging into and replacing elements in an array. If we don't have the ability to identify elements in the array uniquely, we don't know whether a new array element should be added, or the old one should be replaced.

So to summarize from recent comments, we have 3 problems: 
1. We need to be able to merge into all levels of an object, not just arrays (which the merge tool I made, deep-object.js should support).
2. We need to be able to update multiple arrays at a time, ensuring that each one has a key that helps with figuring out whether something should be merged are added.
3. We should be able to pick between removing elements that don't exist in the new array and leaving them alone. Sometimes, like with the array of licenses, we want to remove any that don't exist, and other times, like with the array of releases, we want to just update or add a new one.

Before I start coming up with a solution, I think we should get our test suite added to the project. It will really help make this kind of development easier, as we can do test-driven dev. PR #687 has it setup, but I'm running into some minor dependency issues with Travis' move to Ubuntu 16.04. Should be fixed soon.

**Comment by j-rewerts** [comment]
Oh and 1 more note!

The reason I even went through this pain in the butt way of merging files is because I couldn't figure out how to sort the output of `jsyaml.dump`. If that output the properties and arrays in a deterministic way, we wouldn't need to worry about any of these issues, as you could just replace the entire object and it would show up in GitHub as expected.

**Comment by fredericlestage-vdm** [comment]
From the testing I did, and the many updates I went through to reduce the amount of changes in commits that weren't really changes since it was mostly formatting issues (line length or position of data in the file), I discovered that `jsyaml.dump()` doesn't sort the data at all. It simply takes it as it is in the objects we declare in the `getSomethingObject()` functions. The reason we might have thought that it was sorting in some way might be because we set up the objects in alphabetical order, but its not necessarily showing that way because we then separated mandatory content from optional content, so we end up with two set of alphabetical order. I'm not sure if we could force a totally alphabetical order because when we add new properties to the object (the optional ones), they get added to the end of the object (that already contains all the mandatory properties). Unless we can force the order where we append the property, we will continue to have a semi-alphabetical sorting. The large diffs that showed up with no real change were probably due to the fact that when the data was first created in the yaml format at the beginning of the project, all fields were placed in alphabetic order, but the JS code isn't following that logic (because of the aforementioned separation between mandatory and optional).

TL:DR, `jsyaml.dump()` do output the properties and arrays in a deterministic way, following the structure of the object declared in the `getSomethingObject()` functions.

**Comment by fredericlestage-vdm** [comment]
Resolved by #761
